api/src/main/java/org/openmrs/module/fgh/mpi/miscellaneous/MpiContext.java

```diff
@@ -0,0 +1,263 @@
+package org.openmrs.module.fgh.mpi.miscellaneous;
+
+import static org.openmrs.module.fgh.mpi.MpiConstants.GP_AUTHENTICATION_TYPE;
+import static org.openmrs.module.fgh.mpi.MpiConstants.GP_KEYSTORE_PASS;
+import static org.openmrs.module.fgh.mpi.MpiConstants.GP_KEYSTORE_PATH;
+import static org.openmrs.module.fgh.mpi.MpiConstants.GP_KEYSTORE_TYPE;
+import static org.openmrs.module.fgh.mpi.MpiConstants.GP_MPI_APP_CONTENT_TYPE;
+import static org.openmrs.module.fgh.mpi.MpiConstants.GP_MPI_BASE_URL;
+import static org.openmrs.module.fgh.mpi.MpiConstants.GP_MPI_SYSTEM;
+import static org.openmrs.module.fgh.mpi.MpiConstants.GP_SANTE_CLIENT_ID;
+import static org.openmrs.module.fgh.mpi.MpiConstants.GP_SANTE_CLIENT_SECRET;
+import static org.openmrs.module.fgh.mpi.MpiConstants.GP_UUID_SYSTEM;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+
+import org.apache.commons.lang3.StringUtils;
+import org.openmrs.api.APIException;
+import org.openmrs.api.AdministrationService;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.fgh.mpi.MpiUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MpiContext {
+	
+	private static final Logger log = LoggerFactory.getLogger(MpiContext.class);
+	
+	private TokenInfo tokenInfo;
+	
+	private String contentType;
+	
+	private AuthenticationType authenticationType;
+	
+	private boolean contextInitialized;
+	
+	private SSLContext sslContext;
+	
+	private String serverBaseUrl;
+	
+	private MpiSystemType mpiSystem;
+	
+	private String openmrsUuidSystem;
+	
+	private String clientId;
+	
+	private String clientSecret;
+	
+	public static MpiContext mpiContext;
+	
+	public MpiContext() {
+	}
+	
+	public static synchronized MpiContext initIfNecessary() throws Exception {
+		if (mpiContext == null) {
+			mpiContext = new MpiContext();
+			
+			mpiContext.init();
+		}
+		
+		return mpiContext;
+	}
+	
+	private void init() throws Exception {
```

Comment:
E.g add tests for logic in this method, you can test it indirectly by add tests for `initIfNecessary` if you want to keep this method private

Reply:
Done